api/src/main/java/org/pmiops/workbench/api/WorkspacesController.java

```diff
@@ -690,13 +695,86 @@ private FirecloudWorkspaceId generateFirecloudWorkspaceId(String namespace, Stri
     ReadOnlyNotebookResponse response =
         new ReadOnlyNotebookResponse()
-            .html(notebooksService.getReadOnlyHtml(workspace, workspaceName, notebookName));
+            .html(
+                notebooksService.getReadOnlyHtml(workspaceNamespace, workspaceName, notebookName));
+    return ResponseEntity.ok(response);
+  }
+
+  @Override
+  public ResponseEntity<NotebookLockingMetadataResponse> notebookLockingMetadata(
+      String workspaceNamespace, String workspaceName, String notebookName) {
+
+    // Retrieving the workspace is done first, which acts as an access check.
+    String bucketName =
+        fireCloudService
+            .getWorkspace(workspaceNamespace, workspaceName)
+            .getWorkspace()
+            .getBucketName();
+
+    // response may be empty - fill in what we can
+    NotebookLockingMetadataResponse response = new NotebookLockingMetadataResponse();
+
+    // throws NotFoundException if the notebook is not in GCS
+    // returns null if found but no user-metadata
+    Map<String, String> metadata =
+        cloudStorageService.getMetadata(bucketName, "notebooks/" + notebookName);
+
+    if (metadata != null) {
+      String lockExpirationTime = metadata.get("lockExpirationTime");
+      if (lockExpirationTime != null) {
+        response.lockExpirationTime(Long.valueOf(lockExpirationTime));
+      }
+
+      // stored as a SHA-256 hash of bucketName:userEmail
+      String lastLockedByHash = metadata.get("lastLockedBy");
+      if (lastLockedByHash != null) {
+
+        // the caller should not necessarily know the identities of all notebook users
+        // so we check against the set of users of this workspace which are known to the caller
+
+        // NOTE: currently, users of workspace X of any access level can see all other
+        // workspace X users. This is not desired.
+        // https://precisionmedicineinitiative.atlassian.net/browse/RW-3094
+
+        Set<String> workspaceUsers =
+            workspaceService.getFirecloudWorkspaceAcls(workspaceNamespace, workspaceName).keySet();
+
+        response.lastLockedBy(findHashedUser(bucketName, workspaceUsers, lastLockedByHash));
+      }
+    }
+
     return ResponseEntity.ok(response);
   }
 
+  private String findHashedUser(String bucket, Set<String> workspaceUsers, String hash) {
+    return workspaceUsers.stream()
+        .filter(email -> notebookLockingEmailHash(bucket, email).equals(hash))
+        .findAny()
+        .orElse("UNKNOWN");
+  }
+
+  @VisibleForTesting
+  static String notebookLockingEmailHash(String bucket, String email) {
+    String toHash = String.format("%s:%s", bucket, email);
+    try {
+      MessageDigest sha256 = MessageDigest.getInstance("SHA-256");
+      byte[] hash = sha256.digest(toHash.getBytes(Charset.forName("UTF-8")));
+
+      // convert to printable hex text
+      StringBuilder sb = new StringBuilder();
+      for (byte b : hash) {
+        sb.append(String.format("%02x", b));
```

Comment:
String.format is pretty slow and this would need to run 32*(num users in the acl); since we already pull in Guava - I might prefer an approach like this: https://stackoverflow.com/a/36477814

Reply:
Thanks, updated